plugin-server/src/cdp/consumers/cdp-behavioural-events.consumer.ts

```diff
@@ -82,13 +89,36 @@ export class CdpBehaviouralEventsConsumer extends CdpConsumerBase {
             // Track events consumed and matched (absolute numbers)
             let eventsMatched = 0
             const counterUpdates: CounterUpdate[] = []
+            const occurrenceUpdates: OccurrenceUpdate[] = []
 
             const results = await Promise.all(events.map((event) => this.processEvent(event, counterUpdates)))
             eventsMatched = results.reduce((sum, count) => sum + count, 0)
 
-            // Batch write all counter updates
+            // Prepare occurrence updates for all events, deduplicating within the batch
+            const occurrenceSet = new Set<string>()
+            events.forEach((event) => {
+                const key = `${event.teamId}:${event.personId}:${event.filterGlobals.event}`
+                if (!occurrenceSet.has(key)) {
+                    occurrenceSet.add(key)
+                    occurrenceUpdates.push({
+                        teamId: event.teamId,
+                        personId: event.personId,
+                        eventName: event.filterGlobals.event,
+                    })
+                }
+            })
+
+            // Batch write all updates in parallel
+            const writePromises = []
             if (counterUpdates.length > 0 && this.hub.WRITE_BEHAVIOURAL_COUNTERS_TO_CASSANDRA && this.cassandra) {
-                await this.writeBehavioralCounters(counterUpdates)
+                writePromises.push(this.writeBehavioralCounters(counterUpdates))
+            }
+            if (occurrenceUpdates.length > 0 && this.hub.WRITE_BEHAVIOURAL_COUNTERS_TO_CASSANDRA && this.cassandra) {
+                writePromises.push(this.writePersonEventOccurrences(occurrenceUpdates))
```

Comment:
Thinking ahead to when we actually want to release this - we're going to be doing a _lot_ of writes to cassandra this way right? Without any in memory-caching, if I sent 1000 messages in different batches that will account to 1000s of writes that basically do nothing...

Might need to rethink cassandra as the choice here, or at least the delivery methid